python-lib/tagger_formatter.py

```diff
@@ -0,0 +1,443 @@
+        return tag_columns_in_document, string_sentence, string_keywords
+
+    def _get_tags_row_per_document_category(
+        self, match: Span, line: dict, line_full: dict, sentence: Doc
+    ):
+        """
+        Called by _row_per_document_category
+        Writes the needed informations about founded tags:
+        -line is a dictionary {category:tag}
+        -line_full is a dictionary containing full information about the founded tags
+        """
+        keyword = match.text
+        tag = self.tagger.keyword_to_tag[keyword]
+        category = match.label_
+        sentence = sentence.text
+        if tag not in line_full[category]:
+            json_dictionary = {
+                "occurence": 1,
+                "sentences": [sentence],
+                "keywords": [keyword],
+            }
+            line_full[category][tag] = json_dictionary
+            line[category].append(tag)
+        else:
+            line_full[category][tag]["occurence"] += 1
+            line_full[category][tag]["sentences"].append(sentence)
+            line_full[category][tag]["keywords"].append(keyword)
+        return line, line_full
+
+    def _get_tags_row_per_keyword(self, matches: List, row: pd.Series):
+        """
+        Called by _row_per_keyword
+        Creates the list of new rows with infos about the tags and gives it to _update_output_df function
+        """
+        values = []
+        for match, sentence in matches:
+            values = [
+                {
+                    "tag_keyword": keyword.text,
+                    "tag_sentence": sentence.text,
+                    "tag": self.tagger.keyword_to_tag[keyword.text],
+                }
+                for keyword in match
+            ]
+            self._update_output_df(match, values, row)
+
+    def _get_tags_row_per_keyword_category(self, document: List, row: pd.Series):
+        """
+        Called by _row_per_keyword_category
+        Creates the list of new rows with infos about the tags and gives it to _update_output_df function
+        """
+        tag_rows = []
+        for sentence in document:
+            tag_rows = [
+                {
+                    "tag_keyword": keyword.text,
+                    "tag_sentence": sentence.text,
+                    "tag_category": keyword.label_,
+                    "tag": self.tagger.keyword_to_tag[keyword.text],
+                }
+                for keyword in sentence.ents
+            ]
+            self._update_output_df(tag_rows, tag_rows, row)
+
+    def _update_output_df(self, match: List, values: List[dict], row: pd.Series):
+        """
+        Called by _get_tags_row_per_keyword_category and _get_tags_row_per_keyword
+        Appends:
+        -row with infos about the founded tags to output_df
+        -duplicated initial row from the Document dataframe(text_df) to df.duplicated_lines
+        """
+        if match:
+            self.output_df = self.output_df.append(values, ignore_index=True)
```

Comment:
So we are creating a duplicate of the dataframe all the time?

Reply:
are you talking about the docstring?

Reply:
I am talking about output_df and duplicated_lines